Code/Framework/AzCore/AzCore/Jobs/JobExecutor.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/Jobs/Internal/JobTypeEraser.h>
+#include <AzCore/Jobs/JobDescriptor.h>
+#include <AzCore/std/containers/unordered_map.h>
+#include <AzCore/std/containers/vector.h>
+#include <AzCore/std/parallel/atomic.h>
+#include <AzCore/Memory/PoolAllocator.h>
+
+namespace AZ
+{
+    class JobGraphEvent;
+
+    namespace Internal
+    {
+        class CompiledJobGraph final
+        {
+        public:
+            AZ_CLASS_ALLOCATOR(CompiledJobGraph, SystemAllocator, 0)
+
+            CompiledJobGraph(
+                AZStd::vector<TypeErasedJob>&& jobs,
+                AZStd::unordered_map<uint32_t, AZStd::vector<uint32_t>>& links,
+                size_t linkCount,
+                bool retained);
+
+            ~CompiledJobGraph();
+
+            AZStd::vector<TypeErasedJob>& Jobs() noexcept
+            {
+                return m_jobs;
+            }
+
+            // Indicate that a constituent job has finished and decrement a counter to determine if the
+            // graph should be freed
+            void Release();
+
+        private:
+            friend class JobGraph;
+            friend class JobWorker;
+
+            AZStd::vector<TypeErasedJob> m_jobs;
+            AZStd::vector<uint32_t> m_successors;
+            AZStd::atomic<uint32_t>* m_dependencyCounts = nullptr;
+            JobGraphEvent* m_waitEvent = nullptr;
+            AZStd::atomic<uint32_t> m_remaining;
```

Comment:
👍 I had intended on revisiting this actually. Instead of storing a counter of leaf jobs, I was going to store a counter for job graphs and just increment/decrement wholesale

Reply:
Quick update on this comment: I was on mobile and actually thought you were referring to a different counter (the one tracking outstanding jobs queued on the executor itself).

Padding this to its own cache line makes sense, especially since this counter is allocated at graph granularity. For tracking leaves only though (which is a great suggestion), I was thinking that instrumentation down the line would likely want to quickly get this information. Maybe an optimization for our back pocket? :) The first optimization that shows up in the profiler is any allocation, so I think improving perf for the dynamic dispatch case will likely come first.